File: asonicOFDM/app/src/main/java/com/example/asonicofdm/OFDM.java
package com.example.asonicofdm;

public class OFDM {
    public static double[] modulate(OFDMConfig config, int[] bits) {
        // 目前只写了QPSK
        Complex[] complex_data = Utils.qpsk(bits, config.d);

        // IFFT
        double[][] time_mat_cp = new double[config.symbol_per_carrier][config.IFFT_length + config.GI + config.GIP];
        for (int i = 0; i < config.symbol_per_carrier; i++)
            for (int j = 0; j < config.IFFT_length + config.GI + config.GIP; j++) time_mat_cp[0][0] = 0;

        for (int sym_cnt = 0; sym_cnt < config.symbol_per_carrier; sym_cnt++) {

            // 分别计算ifft
            Complex[] ifft_mod = new Complex[config.IFFT_length];
            double[] ifft_res = new double[config.IFFT_length];
            for (int i = 0; i < config.IFFT_length; i++) ifft_mod[i] = new Complex(0,0);

            for (int car_cnt = 0; car_cnt < config.carrier_count; car_cnt++) {
                ifft_mod[config.carriers[car_cnt]] = complex_data[sym_cnt * config.carrier_count + car_cnt];
                ifft_mod[config.conjugate_carriers[car_cnt]] =
                        (complex_data[sym_cnt * config.carrier_count + car_cnt]).conjugate();
            }
            Complex[] tmp = FFT.ifft(ifft_mod);
            for (int i = 0; i < tmp.length; i++) ifft_res[i] = tmp[i].re();

            // 将结果放入矩阵 添加循环前后缀
            for (int i = 0; i < config.GI; i++) time_mat_cp[sym_cnt][i] = ifft_res[i + config.IFFT_length - config.GI];
            for (int i = 0; i < config.IFFT_length; i++) time_mat_cp[sym_cnt][i + config.GI] = ifft_res[i];
            for (int i = 0; i < config.GIP; i++) time_mat_cp[sym_cnt][config.IFFT_length + config.GI + i] = ifft_res[i];
        }

        // 加窗
        for (int sym_cnt = 0; sym_cnt < config.symbol_per_carrier; sym_cnt++) {
            double[] rcosw = Utils.rcoswindow(config.beta, config.IFFT_length + config.GI);
            for (int i = 0; i < config.IFFT_length + config.GI + 1; i++)
                time_mat_cp[sym_cnt][i] *= rcosw[i];
        }

        // 并转串
        double[] Tx_data = new double[config.symbol_per_carrier*(config.IFFT_length + config.GI)+config.GIP];
        for (int i = 0; i < config.IFFT_length + config.GI + config.GIP; i++)
            Tx_data[i] = time_mat_cp[0][i];
        for (int i = 1; i < config.symbol_per_carrier; i++)
            // for (int j = (config.IFFT_length + config.GI)*i; j < (config.IFFT_length + config.GI)*(i+1)+config.GIP; j++)
            for (int j = 0; j < config.IFFT_length + config.GI + config.GIP; j++)
                Tx_data[(config.IFFT_length + config.GI)*i + j] = time_mat_cp[i][j];

        return Tx_data;
    }

    public static 
}

File: IQ_version/OFDM.m
function Tx_data = OFDM(config, bits)
%
% Syntax: Tx_data = OFDM(bits)
%

% ======================== 编码映射 ============================
    if config.modulate == 1
        complex_data = QAM16(bits, config.d);
    elseif config.modulate == 2
        complex_data = QPSK(bits, config.d);
    else
        complex_data = BPSK(bits, config.d);
    end
    complex_mat = reshape(complex_data', config.carrier_count, config.symbol_per_carrier)';

% ======================== IFFT ============================
    IFFT_mod = zeros(config.symbol_per_carrier, config.IFFT_length);
    IFFT_mod(:, config.carriers) = complex_mat;
    % IFFT_mod(:, config.conjugate_carriers) = conj(complex_mat);

    IFFT_res = ifft(IFFT_mod, config.IFFT_length, 2);

% ======================== 循环前后缀 ============================
    time_mat_cp = zeros(config.symbol_per_carrier, config.IFFT_length + config.GI + config.GIP);
    for k = 1:config.symbol_per_carrier
        for i = 1:config.GI
            time_mat_cp(k, i) = IFFT_res(k, i + config.IFFT_length - config.GI);
        end
        for i = 1:config.IFFT_length
            time_mat_cp(k, i + config.GI) = IFFT_res(k, i);
        end
        for i = 1:config.GIP
            time_mat_cp(k, config.IFFT_length + config.GI + i) = IFFT_res(k, i);
        end
    end

% ======================== 信号加窗 ==============================
    windowed_time_mat_cp = zeros(config.symbol_per_carrier, config.IFFT_length + config.GI + config.GIP);
    for i = 1:config.symbol_per_carrier
        %加窗 升余弦窗
        windowed_time_mat_cp(i,:) = time_mat_cp(i,:).*rcoswindow(...
            config.beta, config.IFFT_length + config.GI + config.GIP)';
        % Or
        % windowed_time_mat_cp(i,:) = time_mat_cp(i,:).*hann(config.IFFT_length + config.GI + config.GIP)';
    end

% ======================== 并转串 ==============================
    % Tx_data = reshape(time_mat_cp', (config.symbol_per_carrier) * (config.IFFT_length + config.GI),1)';
    windowed_Tx_data = zeros(1,config.symbol_per_carrier*(config.IFFT_length + config.GI)+config.GIP);
    windowed_Tx_data(1:config.IFFT_length + config.GI + config.GIP) = windowed_time_mat_cp(1,:);
    for i = 1:config.symbol_per_carrier - 1;
        windowed_Tx_data((config.IFFT_length + config.GI)*i+1:(...
            config.IFFT_length + config.GI)*(i+1)+config.GIP)=...
            windowed_time_mat_cp(i+1,:);%并串转换，循环后缀与循环前缀相叠加
    end

    Tx_data = windowed_Tx_data;
end
